### PR TITLE
AWS Elasticache package

### PR DESCRIPTION
**Pull request by matteo-magni**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -69,7 +69,8 @@ imgpkg push -b ${REPO_HOST}:${TAG} -f repository
 
 The tanzu-application-platform-reference-service-packages project team welcomes contributions from the community. Before you start working with this project please
```
Comment (Samze) [inline README.md:70]: *nitpick* We should update the README table to include the new elasticcache pacakge

File: bundles/amazon/ack/elasticache/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+1. Build a new Package bundle image:
+
+```
+export REPO_HOST=<YOUR_IMAGE_REPO> #! e.g. ghcr.io/vmware-tanzu/tanzu-application-platform-reference-service-packages
+export BUNDLE_TAG=<YOUR_BUNDLE_TAG> #! e.g. latest
+
+pushd bundle
+    kbld -f config/ --imgpkg-lock-output=.imgpkg/images.yml
+    imgpkg push -b ${REPO_HOST}/psql.aws.references.services.apps.tanzu.vmware.com:$BUNDLE_TAG -f .
```
Comment (Samze) [inline bundles/amazon/ack/elasticache/README.md:54]: *nitpick* This refers to psql

File: bundles/amazon/ack/elasticache/bundle/config/00-schema.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+#@schema/title "Namespace"
+#@schema/desc "Namespace to deploy the kubernetes resources to"
+namespace: ""
+
+#@schema/title "CreateNamespace"
+#@schema/desc "Whether to create the namespace for the resources or not"
+createNamespace: False
+
+#@schema/title "CacheSubnetGroupName"
+#@schema/desc "Name of the cache subnet group to create the replication group in. It has to be available upfront."
```
Comment (Samze) [inline bundles/amazon/ack/elasticache/bundle/config/00-schema.yml:16]: *nitpick*
```suggestion
#@schema/desc "Name of the cache subnet group to create the replication group in. It must be pre-created."
```

File: bundles/amazon/ack/elasticache/bundle/config/00-schema.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+cacheNodeType: "cache.t2.micro"
+
+#@schema/title "VpcSecurityGroupIDs"
+#@schema/desc "The list of security groups to associate to the replication group"
+vpcSecurityGroupIDs:
+  - ""
+
+#@schema/title "EngineVersion"
+#@schema/desc "The version of the Redis engine to use. Available versions can be obtained running `aws elasticache describe-cache-engine-versions --engine redis`."
+#@schema/default "6.2"
```
Comment (Samze) [inline bundles/amazon/ack/elasticache/bundle/config/00-schema.yml:30]: **question** Curious why we use `#@schema/default` here over setting the value? It seems like `cacheNodeType` does the opposite.

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml [hunk 5]
```diff
@@ -0,0 +1,225 @@
+#@ load("@ytt:data", "data")
+#@ load("@ytt:assert", "assert")
+
+#@ def get_id(x): return "{}-{}".format(data.values.name, x)
+
+#@ def tags(resource_tags=[]):
+#@   return [ x for x in (data.values.tags + resource_tags) if x["key"] != "" and x["value"] != "" ]
```
Comment (Samze) [inline bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:7]: **question** Kind of unfortunate we have to special case `key` and `value here. Do we need to set these in the schema?

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+  - name: creds
+    ref:
+      apiVersion: v1
+      kind: Secret
+      name: #@ user_creds
+  template:
+    metadata:
+      labels:
+        services.apps.tanzu.vmware.com/class: redis-small
+    type: servicebinding.io/redis
```
Comment (Samze) [inline bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:101]: **suggestion** Not part of this PR, but it would be great to fix the other bundles we have. Right nwo they do not have the `servicebinding.io` prefix. E.g. https://github.com/vmware-tanzu/tanzu-application-platform-reference-service-packages/blob/main/bundles/amazon/ack/rds/bundle/config/config.yaml#L75

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+      name: #@ data.values.name
+  - name: creds
+    ref:
+      apiVersion: v1
+      kind: Secret
+      name: #@ user_creds
+  template:
+    metadata:
+      labels:
+        services.apps.tanzu.vmware.com/class: redis-small
```
Comment (Samze) [inline bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:100]: **suggestion** I think we should expose this as a parameter so its configurable. We did this for the Google bundle https://github.com/vmware-tanzu/tanzu-application-platform-reference-service-packages/blob/main/bundles/google/config-connector/cloudsql/bundle/config/00-schema.yml#L33

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    type: servicebinding.io/redis
+    stringData:
+      type: redis
+      username: #@ user_name
+      ssl: "true"
+#@ if data.values.clusterMode:
+      host: $(.replicationGroup.status.configurationEndpoint.address)
```
Comment (Samze) [inline bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:107]: **question** Do we want to set the `cluster.nodes` property here as defined per https://github.com/spring-cloud/spring-cloud-bindings#redis we can then provide the address/port of all nodes?

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+- apiGroups:
+  - elasticache.services.k8s.aws
+  resources:
+  - replicationgroups
+  verbs:
+  - get
+  - list
+  - watch
+  resourceNames:
+  - #@ data.values.name
```
Comment (Samze) [inline bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:211]: **nitpick** I know they are the same right now but using the var makes sense

```suggestion
  - #@ replicationGroupName
```

File: bundles/amazon/ack/elasticache/bundle/config/00-schema.yml (hunk 4 above)
Comment (matteo-magni) [reply bundles/amazon/ack/elasticache/bundle/config/00-schema.yml:30]: You're right, we don't need `#@schema/default` annotations for strings, numbers and booleans

File: bundles/amazon/ack/elasticache/README.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
+# AWS Controllers for Kubernetes - Elasticache Redis Instance
+
+Status: Experimental
+
+## Use
+
+Create a `CacheSubnetGroup`, for example:
```
Comment (teddyking) [inline bundles/amazon/ack/elasticache/README.md:7]: **question** Could/should we create this as part of the package install?

File: bundles/amazon/ack/elasticache/bundle/config/00-schema.yml (hunk 3 above)
Comment (matteo-magni) [reply bundles/amazon/ack/elasticache/bundle/config/00-schema.yml:16]: As non-English native speaker, I accept all kinds of suggestions like this one 🙂

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml (hunk 5 above)
Comment (matteo-magni) [reply bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:7]: Sorry, I didn't get the question. What do you mean?

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml (hunk 6 above)
Comment (matteo-magni) [reply bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:101]: Correct. We also have to fix the folder names, like making `amazon` become `aws`. 
We'll have a separate piece of work and PR for this.

File: bundles/amazon/ack/elasticache/README.md (hunk 10 above)
Comment (matteo-magni) [reply bundles/amazon/ack/elasticache/README.md:7]: We _could_, but that would mean providing even more ephemeral values to the package. We're currently accepting a list of Security Group IDs, which are pretty much ephemeral (you destroy and recreate the SGs and you get different IDs, meaning that you'd need to update your installed package), but there's no other way to do it.
If we wanted to create the `CacheSubnetGroup` as part of the package then we'd need to have also the IDs of the subnets to create the group for. Unfortunately there's no such a thing as data sources for Terraform I could use to read the IDs (i.e. matching a tag).
However, we could have also a boolean like `createCacheSubnetGroup` as input value (along with optional subnet IDs) to determine if the user wants the `CacheSubnetGroup` to be created. We just need some more ytt logic but it's doable.

**Comment by teddyking** [comment]
Just tried the `Package` out and got it to deploy successfully. I was able to confirm that an elasticache instance came up in the AWS console and that 2 x bindable Secrets were created - one for the `reader` and one for the `writer` user.

I'm now using the redis-cli to attempt to test the service out but I don't seem to be able to connect to the server ... I have confirmed basic connectivity using netcat, so I can reach the service, however I can't seem to login using the redis-cli and the username/password from the bindings. I wonder if perhaps it's something to do with SSL? Are we missing certs/keys or something? Have you been able to test the service end-to-end?

**Comment by matteo-magni** [comment]
> I'm now using the redis-cli to attempt to test the service out but I don't seem to be able to connect to the server ... I have confirmed basic connectivity using netcat, so I can reach the service, however I can't seem to login using the redis-cli and the username/password from the bindings. I wonder if perhaps it's something to do with SSL? Are we missing certs/keys or something? Have you been able to test the service end-to-end?

It is related to SSL, we've already experienced this problem using the official `redis:6.2` container. First of all you have to force SSL conection using the `--tls` flag, but, as the redis container does not have the `ca-certificates` package installed, it cannot verify the ssl chain. So either you add the `--insecure` flag or you install the above apt package.
The CA that signed the Elasticache instance's certificate is public, therefore you don't need to specify any private certificate for it.

**Comment by teddyking** [comment]
Ah yep that was it - after installing the `ca-certificates` package and providing the `--tls` flag I was able to establish a connection to service instance. Nice! I experimented a bit first by using the `writer` user credentials - I SET and GET a test value and confirmed it worked. I then logged in with the `reader` user credentials and confirmed that I could not SET a test value, but that I could still GET the value.

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml (hunk 8 above)
Comment (matteo-magni) [reply bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:113]: I think I'm going to disable the `clusterMode` feature until [this issue](https://github.com/aws-controllers-k8s/community/issues/1513) is fixed. It's a race condition that's preventing the ReplicationGroup to be created deterministically.

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml (hunk 5 above)
Comment (Samze) [reply bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:8]: Maybe asked another way, why do we need this statement?

```code
if x["key"] != "" and x["value"] != ""
```

File: bundles/amazon/ack/elasticache/README.md (hunk 10 above)
Comment (Samze) [reply bundles/amazon/ack/elasticache/README.md:7]: We also didn't include `dbSubnetGroup` as part of the RDS bundle for similar reasons. Whatever we decide we should ensure we do the same for the RDS bundle too.

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml (hunk 5 above)
Comment (matteo-magni) [reply bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:8]: It's for validating the inputs.
Without that statement the following input
```yaml
tags:
  - value: a value without a key
```
would produce
```yaml
apiVersion: elasticache.services.k8s.aws/v1alpha1
kind: ReplicationGroup
metadata:
  name: trp01
  namespace: elasticache-01
spec:
  description: A redis service instance
  engine: redis
  engineVersion: "6.2"
  replicationGroupID: trp01
  cacheNodeType: cache.t2.micro
  cacheSubnetGroupName: trp
  cacheParameterGroupName: trp01
  securityGroupIDs:
  - sg-0a4ddae4fbf426cc8
  atRestEncryptionEnabled: true
  transitEncryptionEnabled: true
  replicasPerNodeGroup: 0
  userGroupIDs:
  - trp01
  tags:
  - value: a value without a key
    key: ""
```

Comment (Samze) [reply bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:8]: Understood, makes sense 👍

Comment (matteo-magni) [reply bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:8]: Until we have proper input validation built into ytt we have to use workarounds like this..

File: bundles/amazon/ack/elasticache/README.md (hunk 10 above)
Comment (teddyking) [reply bundles/amazon/ack/elasticache/README.md:7]: Cool makes sense - perhaps lets leave it as it is for now then.

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml (hunk 8 above)
Comment (matteo-magni) [reply bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:113]: > **question** Do we want to set the `cluster.nodes` property here as defined per https://github.com/spring-cloud/spring-cloud-bindings#redis when in `clusterMode==true` we can then provide the address/port of all nodes?

@Samze what kind of list is `cluster.nodes` supposed to be holding? Even with cluster mode disabled I can have multiple read replicas. Does `cluster.nodes` need to be set as a comma-separated list of both the primary node and all its read replicas?

File: bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml (hunk 9 above)
Comment (matteo-magni) [reply bundles/amazon/ack/elasticache/bundle/config/01-replication-group.ytt.yml:207]: 👍
